lib/model/message.dart

```diff
@@ -66,6 +234,219 @@ mixin MessageStore {
+  void _updateOutboxMessage(int localMessageId, {
+    required OutboxMessageState newState,
+  }) {
+    assert(!_disposed);
+    final outboxMessage = outboxMessages[localMessageId];
+    if (outboxMessage == null) {
+      throw StateError(
+        'Removing unknown outbox message with localMessageId: $localMessageId');
+    }
+    final oldState = outboxMessage.state;
+    // See [OutboxMessageState] for valid state transitions.
+    final isStateTransitionValid = switch (newState) {
+      OutboxMessageState.hidden => false,
+      OutboxMessageState.waiting =>
+        oldState == OutboxMessageState.hidden,
+      OutboxMessageState.waitPeriodExpired =>
+        oldState == OutboxMessageState.waiting,
+      OutboxMessageState.failed =>
+        oldState == OutboxMessageState.hidden
+        || oldState == OutboxMessageState.waiting
+        || oldState == OutboxMessageState.waitPeriodExpired,
+    };
+    if (!isStateTransitionValid) {
+      throw StateError('Unexpected state transition: $oldState -> $newState');
+    }
+
+    outboxMessage._state = newState;
+    for (final view in _messageListViews) {
+      if (oldState == OutboxMessageState.hidden) {
+        view.addOutboxMessage(outboxMessage);
+      } else {
+        view.notifyListenersIfOutboxMessagePresent(localMessageId);
+      }
+    }
+  }
+
+  /// Send a message and create an entry of [OutboxMessage].
+  Future<void> _outboxSendMessage({
+    required MessageDestination destination,
+    required String content,
+    required String? realmEmptyTopicDisplayName,
+  }) async {
+    assert(!_disposed);
+    final localMessageId = _nextLocalMessageId++;
+    assert(!outboxMessages.containsKey(localMessageId));
+
+    final conversation = switch (destination) {
+      StreamDestination(:final streamId, :final topic) =>
+        StreamConversation(
+          streamId,
+          _processTopicLikeServer(
+            topic, realmEmptyTopicDisplayName: realmEmptyTopicDisplayName),
+          displayRecipient: null),
+      DmDestination(:final userIds) => DmConversation(allRecipientIds: userIds),
+    };
+
+    _outboxMessages[localMessageId] = OutboxMessage.fromConversation(
+      conversation,
+      localMessageId: localMessageId,
+      selfUserId: selfUserId,
+      timestamp: ZulipBinding.instance.utcNow().millisecondsSinceEpoch ~/ 1000,
+      contentMarkdown: content);
+
+    _outboxMessageDebounceTimers[localMessageId] = Timer(
+      kLocalEchoDebounceDuration,
+      () => _handleOutboxDebounce(localMessageId));
+
+    _outboxMessageWaitPeriodTimers[localMessageId] = Timer(
+      kSendMessageOfferRestoreWaitPeriod,
+      () => _handleOutboxWaitPeriodExpired(localMessageId));
+
+    try {
+      await _apiSendMessage(connection,
+        destination: destination,
+        content: content,
+        readBySender: true,
+        queueId: queueId,
+        localId: localMessageId.toString());
+      _outboxMessageWaitPeriodTimers.remove(localMessageId)?.cancel();
```

Comment:
Cool, I guess this line is the implementation of https://github.com/zulip/zulip-flutter/pull/1472#discussion_r2099101759 .